RecoEcal/EgammaCoreTools/src/DeepSCGraphEvaluation.cc

```diff
@@ -0,0 +1,210 @@
+  std::ifstream inputfile{edm::FileInPath(file).fullPath()};
+  int ninputs = 0;
+  if (inputfile.fail()) {
+    throw cms::Exception("MissingFile") << "Scaler file not found: " << file;
+  } else {
+    // Now read mean, scale factors for each variable
+    float par1, par2;
+    while (inputfile >> par1 >> par2) {
+      scalingParams.push_back(std::make_pair(par1, par2));
+      ninputs += 1;
+    }
+  }
+  return ninputs;
+}
+
+std::vector<double> DeepSCGraphEvaluation::scaleClusterFeatures(const std::vector<double>& input) const {
+  std::vector<double> out(input.size());
+  for (size_t i = 0; i < input.size(); i++) {
+    const auto& [par1, par2] = scalerParamsClusters_[i];
+    out[i] = (input[i] - par1) / par2;
+  }
+  return out;
+}
+
+std::vector<double> DeepSCGraphEvaluation::scaleWindowFeatures(const std::vector<double>& input) const {
+  std::vector<double> out(input.size());
+  for (size_t i = 0; i < input.size(); i++) {
+    const auto& [par1, par2] = scalerParamsWindows_[i];
+    out[i] = (input[i] - par1) / par2;
+  }
+  return out;
+}
+
+std::vector<std::vector<float>> DeepSCGraphEvaluation::evaluate(const DeepSCInputs& inputs) const {
+  /*
+   Evaluate the DeepSC model
+  */
+  LogDebug("DeepSCGraphEvaluation") << "Starting evaluation";
+
+  // Final output
+  std::vector<std::vector<float>> outputs_clustering;
+
+  // We need to split the total inputs in N batches of size batchSize (configured in the producer)
+  // being careful with the last batch which will have less than batchSize elements
+  size_t nInputs = inputs.clustersX.size();
+  uint iB = -1;  // batch index
+  while (nInputs > 0) {
+    iB++;  // go to next batch
+    size_t nItems;
+    if (nInputs >= cfg_.batchSize) {
+      nItems = cfg_.batchSize;
+      nInputs -= cfg_.batchSize;
+    } else {
+      nItems = nInputs;
+      nInputs = 0;
+    }
+    // Input tensors initialization
+
+    // Inputs
+    tensorflow::Tensor clsX_{tensorflow::DT_FLOAT,
+                             {static_cast<long int>(nItems), cfg_.maxNClusters, cfg_.nClusterFeatures}};
+    tensorflow::Tensor windX_{tensorflow::DT_FLOAT, {static_cast<long int>(nItems), cfg_.nWindowFeatures}};
+    tensorflow::Tensor hitsX_{
+        tensorflow::DT_FLOAT,
+        {static_cast<long int>(nItems), cfg_.maxNClusters, cfg_.maxNRechits, cfg_.nRechitsFeatures}};
+    tensorflow::Tensor isSeedX_{tensorflow::DT_FLOAT, {static_cast<long int>(nItems), cfg_.maxNClusters, 1}};
+    tensorflow::Tensor nClsSize_{tensorflow::DT_FLOAT, {static_cast<long int>(nItems)}};
+
+    float* C = clsX_.flat<float>().data();
+    // Look on batch dim
+    for (size_t b = 0; b < nItems; b++) {
+      const auto& cls_data = inputs.clustersX[iB * cfg_.batchSize + b];
+      // Loop on clusters
+      for (size_t k = 0; k < cfg_.maxNClusters; k++) {
+        // Loop on features
+        for (size_t z = 0; z < cfg_.nClusterFeatures; z++, C++) {
+          if (k < cls_data.size()) {
+            *C = float(cls_data[k][z]);
```

Comment:
I'm aware of the `input.matrix<float>()(i,j)` syntax, but I opted for the direct manipulation to avoid any overhead as explained in https://cms-ml.github.io/documentation/inference/tensorflow2.html#tensor-data-access-via-pointers. If you prefer the other way for readability I can change it.

Reply:
do you know how much overhead there is from this (compared to the model itself)? From what I've seen, direct pointer manipulation can be somewhat hard to maintain in case something changes in the features later.